tools/prow-auto-bumper/getversion_test.go

```diff
@@ -0,0 +1,368 @@
+		}
+
+		pv.parseChangelist(fc)
+		if eq := reflect.DeepEqual(pv.images, data.images); !eq {
+			t.Fatalf("parsing PR with changes '%v', want: '%v', got: '%v'",
+				data.patches, data.images, pv.images)
+		}
+	}
+}
+
+func TestGetBestVersion(t *testing.T) {
+	type PRInfo struct {
+		delta      int // hours before current date
+		state      ghutil.PullRequestState
+		oldVersion string
+		newVersion string
+	}
+	datas := []struct {
+		PIs        []PRInfo
+		dominantVs *versions
+	}{
+		{
+			[]PRInfo{
+				{7 * 24, ghutil.PullRequestCloseState, "va-b", "vh-i"},
+			},
+			&versions{"va-b", "vh-i", ""},
+		},
+		{
+			[]PRInfo{
+				{7 * 24, ghutil.PullRequestCloseState, "va-b", "vh-i"},
+				{6 * 24, ghutil.PullRequestCloseState, "vc-d", "vj-k"},
+				{8 * 24, ghutil.PullRequestCloseState, "ve-f", "vl-m"},
+			},
+			&versions{"va-b", "vh-i", ""},
+		},
+		{ // Reverted
+			[]PRInfo{
+				{7 * 24, ghutil.PullRequestCloseState, "va-b", "vh-i"},
+				{7*24 - 3, ghutil.PullRequestCloseState, "vh-i", "va-b"}, // later PR
+			},
+			&versions{"vh-i", "va-b", ""},
+		},
+		{
+			[]PRInfo{
+				{7 * 24, ghutil.PullRequestOpenState, "va-b", "vh-i"},
+			},
+			nil,
+		},
+		{
+			[]PRInfo{
+				{4 * 24, ghutil.PullRequestCloseState, "va-b", "vh-i"},
+				{10 * 24, ghutil.PullRequestCloseState, "vc-d", "vj-k"},
+			},
+			nil,
+		},
+	}
+
+	for _, data := range datas {
+		fgc := fakeghutil.NewFakeGithubClient()
+		fc := &Client{fgc}
+		dataNow := time.Now()
+		for i, PI := range data.PIs {
+			PR, err := fgc.CreatePullRequest(fakeOrg, fakeRepo, "user:head", "base", "title", "body")
+			if nil != err {
+				t.Fatalf("Create PR in %s/%s, want: no error, got: '%v'", fakeOrg, fakeRepo, err)
+			}
+			timeCreated := dataNow.Add(-time.Hour * time.Duration(PI.delta))
+			stateStr := string(PI.state)
+			PR.State = &stateStr
+			PR.CreatedAt = &timeCreated
+			SHA := strconv.Itoa(i)
+			filename := "fakefile"
+			patch := fmt.Sprintf(`
+				- image: gcr.io/k8s-foofoo/bar:%s
+				+ image: gcr.io/k8s-foofoo/bar:%s
+			`, PI.oldVersion, PI.newVersion)
+			fgc.AddCommitToPullRequest(fakeOrg, fakeRepo, *PR.Number, SHA)
+			fgc.AddFileToCommit(fakeOrg, fakeRepo, SHA, filename, patch)
+		}
+
+		pv, err := getBestVersion(fc, fakeOrg, fakeRepo, "user:head", "base")
+		if nil != err {
+			t.Fatalf("get best versions with PRs '%v', want: no error, got: '%v'", data.PIs, err)
+		}
+		if nil == data.dominantVs {
+			if nil != pv && nil != pv.dominantVs {
+				t.Fatalf("get best versions with PRs '%v', want: nil, got: '%v'", data.PIs, pv.getDominantVersions())
+			}
+		} else if eq := reflect.DeepEqual(*data.dominantVs, pv.getDominantVersions()); !eq {
+			t.Fatalf("get best versions with PRs '%v', want: '%v', got: '%v'", data.PIs, data.dominantVs, pv.getDominantVersions())
+		}
+	}
+}
+
+func TestRetryGetBestVersion(t *testing.T) {
+	fgc := fakeghutil.NewFakeGithubClient()
+	fc := &Client{fgc}
+
+	_, err := retryGetBestVersion(fc, fakeOrg, fakeRepo, "user:head", "base")
```

Comment:
That's a funny test. :) Why you expect it to always fail?

Reply:
Only failure can make it exercise all code in the function :)